compiler/rustc_query_impl/src/plumbing.rs

```diff
@@ -409,53 +477,14 @@ macro_rules! define_queries {
             $(pub(crate) fn $name()-> DepKindStruct {
```

Comment:
Can this whole function be a single call to `query_callback` with `queries::$name`, `is_anon` and `is_eval_always` given as arguments?

Reply:
No, unfortunately. We need to at least pass in `TyCtxt::$name`, because nothing associates it to `queries::$name`. And we can't get rid of the closure calling `force_from_dep_node` because of the same variance issue as above:
```
error[E0308]: mismatched types
   --> compiler/rustc_query_impl/src/plumbing.rs:351:35
    |
351 |         force_from_dep_node: Some(force_from_dep_node::<Q>),
    |                              ---- ^^^^^^^^^^^^^^^^^^^^^^^^ one type is more general than the other
    |                              |
    |                              arguments to this enum variant are incorrect
    |
    = note: expected fn pointer `for<'r> fn(rustc_middle::ty::TyCtxt<'r>, DepNode<_>) -> _`
                  found fn item `fn(rustc_middle::ty::TyCtxt<'_>, DepNode<_>) -> _ {force_from_dep_node::<'_, Q>}`
```

Reply:
Ok, I think I can make this work by parameterizing `DepKindStruct` over a lifetime instead of requiring an HRTB for the function pointers it stores. Working on that now.